tools/smartdnsblock/smartdnsblock/smartdnsblock.cpp

```diff
@@ -0,0 +1,135 @@
+// Copyright 2018 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// If this doesn't come first, there will be compile errors.
+#include <winsock2.h>
+
+#include <iostream>
+
+#include <iphlpapi.h>
+
+#include <fwpmtypes.h>
+#include <fwpmu.h>
+
+#pragma comment(lib, "iphlpapi.lib")
+#pragma comment(lib, "fwpuclnt.lib")
+
+using namespace ::std;
+
+PCWSTR TAP_DEVICE_NAME = L"outline-tap0";
+ULONG GET_ADAPTERS_ADDRESSES_BUFFER_SIZE = 16384;
+
+PCWSTR FILTER_PROVIDER_NAME = L"Outline";
+
+int main(int argc, char **argv) {
+  // Lookup the interface index of outline-tap0.
+  PIP_ADAPTER_ADDRESSES adaptersAddresses =
+      (IP_ADAPTER_ADDRESSES *)malloc(GET_ADAPTERS_ADDRESSES_BUFFER_SIZE);
+  DWORD result = GetAdaptersAddresses(AF_INET, 0, NULL, adaptersAddresses,
+                                      &GET_ADAPTERS_ADDRESSES_BUFFER_SIZE);
+  if (result != NO_ERROR) {
+    cerr << "could not fetch network device list: " << result << endl;
+    return 1;
+  }
+
+  UINT32 interfaceIndex;
+  PIP_ADAPTER_ADDRESSES adapterAddress = adaptersAddresses;
+  while (adapterAddress && wcscmp(TAP_DEVICE_NAME, adapterAddress->FriendlyName) != 0) {
+    adapterAddress = adapterAddress->Next;
+  }
+
+  if (!adapterAddress) {
+    wcerr << "could not find " << TAP_DEVICE_NAME << endl;
```

Comment:
nit: why do you sometimes use `cerr/cout` and others `wcerr/wcout`?

Reply:
`werr` and `wcout` for wide character support, necessary since the WFP APIs expect names in wide character format. I agree it's weird to use a mix of `cout` and `wcout` so I just changed it to use the wide character variants throughout.